org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinInlineFilter.java

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.LineNumberNode;
+import org.objectweb.asm.tree.MethodNode;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Filters out instructions that were inlined by Kotlin compiler.
+ */
+public final class KotlinInlineFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterContext context, final IFilterOutput output) {
+		if (context.getSourceDebugExtension() == null) {
+			return;
+		}
+		if (!context.getClassAnnotations()
+				.contains(KotlinGeneratedFilter.KOTLIN_METADATA_DESC)) {
+			return;
+		}
+
+		final int max = parseSMAP(context.getSourceDebugExtension());
+		int line = 0;
+		for (AbstractInsnNode i = methodNode.instructions
+				.getFirst(); i != null; i = i.getNext()) {
+			if (AbstractInsnNode.LINE == i.getType()) {
+				line = ((LineNumberNode) i).line;
+			}
+			if (line >= max) {
+				output.ignore(i, i);
+			}
+		}
+	}
+
+	private static int parseSMAP(final String smap) {
```

Comment:
Took me some time to understand the exclusion logic. Maybe renaming this method to `getFirstIncludeLineNumber()` or so would help to understand how the references to external files are implemented. Also a link to Kotlin documentation or a short description of the logic in the filter's JavaDoc might be helpful for future reference.

Reply:
So far I haven't found any spec that at least states that Kotlin uses SMAP, not even talking about additional `KotlinDebug` strata. And maybe there is none at all. Understanding of how to parse comes from JSR-45 and how to interpret from `kotlinc`. But anyway you right - will work on addition of description.

Reply:
Renamed into `getFirstGeneratedLineNumber`